service/matching/version_rule_helpers.go

```diff
@@ -515,3 +539,154 @@ func calcRampThreshold(id string) float64 {
 	h := farm.Fingerprint32([]byte(id))
 	return 100 * (float64(h) / (float64(math.MaxUint32) + 1))
 }
+
+/*
+e.g.
+Redirect Rules:
+10
+^
+|
+1 <------ 2
+^
+|
+5 <------ 3 <------ 4
+
+Assignment Rules:
+[ (3, 50%), (2, nil) ]
+*/
+func getUpstreamBuildIds(
```

Comment:
once we include the recently-deleted redirect rules, there is no guarantee of acyclic graph. please make sure this function does not fall into an infinite loop in that case.